packages/client/src/Compose/RecipientsInput.tsx

```diff
@@ -67,46 +101,130 @@ function reducer(state: State, action: Action): State {
     recipients: [],
     inputValue: ""
   })
+
   const [focused, setFocused] = React.useState(false)
 
+  const { data } = graphql.useGetMatchingAddressesQuery({
+    variables: { inputValue }
+  })
+
+  let suggestions = data ? data.addresses : []
+
+  const skip = inputValue.trim().length < 3
+
+  function renderSuggestion(
+    suggestion: graphql.Address,
+    { query, isHighlighted }: Autosuggest.RenderSuggestionParams
+  ) {
+    const matches = match(getSuggestionValue(suggestion), query)
+    const results = parse(getSuggestionValue(suggestion), matches)
+
+    return (
+      <MenuItem selected={isHighlighted} component="div">
+        <div>
+          {results.map(result => (
+            <span
+              key={result.text}
+              style={{ fontWeight: result.highlight ? 500 : 400 }}
+            >
+              {result.text}
+            </span>
+          ))}
+        </div>
+      </MenuItem>
+    )
+  }
+
+  function getSuggestionValue(suggestion: graphql.Address) {
+    return suggestion.name
+      ? `${suggestion.name} <${suggestion.mailbox}@${suggestion.host}>`
+      : `${suggestion.mailbox}@${suggestion.host}`
+  }
+
+  const handleSuggestionsFetchRequested = () => {
+    return suggestions
+  }
+
+  const handleSuggestionsClearRequested = () => {
+    suggestions = []
+  }
+
+  const onSuggestionSelected = () => {
+    dispatch({ type: "parseAddresses" })
+  }
+
+  function renderInputComponent(
+    inputProps: Autosuggest.InputProps<typeof suggestions[number]>
+  ) {
+    return (
+      <TextField
+        {...(inputProps as any)}
+        {...rest}
+        fullWidth
+        InputProps={{
+          startAdornment: (
+            <InputAdornment position="start">
+              {recipients.map(recipient => (
+                <Chip
+                  key={email(recipient)}
+                  avatar={
+                    <Avatar
+                      address={{
+                        name: recipient.name,
+                        mailbox: recipient.local,
+                        host: recipient.domain
+                      }}
+                    />
+                  }
+                  tabIndex={-1}
+                  label={display(recipient)}
+                  className={classes.chip}
+                  onDelete={() => dispatch({ type: "remove", recipient })}
+                />
+              ))}
+            </InputAdornment>
+          )
+        }}
+        InputLabelProps={{
+          shrink: recipients.length > 0 || inputValue.length > 0 || focused
+        }}
+      />
+    )
+  }
+
+  const autosuggestProps = {
+    renderInputComponent,
+    suggestions: skip ? [] : suggestions,
+    onSuggestionsFetchRequested: handleSuggestionsFetchRequested,
+    onSuggestionsClearRequested: handleSuggestionsClearRequested,
+    onSuggestionSelected,
+    getSuggestionValue,
+    renderSuggestion
+  }
+
   React.useEffect(() => {
```

Comment:
Please group hooks at the top of the component. It is important that every hook runs every time the component renders; if a hook is called too far down we might accidentally put in an early return or something that could break things. I also like having hooks at the top because they inform someone reading the code of important side-effects, and it is useful to be able to look at those details in one place.